<h1 align= 'center'><b>INX Future Inc Employee Performance</h1></b>

#### **BUSINESS CASE**

INX Future Inc , (referred as INX ) , is one of the leading data analytics and automation solutions provider 
with over 15 years of global business presence. INX is consistently rated as top 20 best employers past 5 
years. INX human resource policies are considered as employee friendly and widely perceived as best 
practices in the industry. 

Recent years, the employee performance indexes are not healthy and this is becoming a growing 
concerns among the top management. There has been increased escalations on service delivery and 
client satisfaction levels came down by 8 percentage points. 

CEO, Mr. Brain, knows the issues but concerned to take any actions in penalizing non-performing 
employees as this would affect the employee morale of all the employees in general and may further 
reduce the performance. Also, the market perception best employer and thereby attracting best talents 
to join the company.  

Mr. Brain decided to initiate a data science project , which analyses the current employee data and find 
the core underlying causes of this performance issues. Mr. Brain, being a data scientist himself, expects 
the findings of this project will help him to take right course of actions. He also expects a clear indicators 
of non performing employees, so that any penalization  of non-performing employee, if required, may 
not significantly affect other employee morals.

#### **PROBLEM STATEMENT**

- To develop a data science project to identify the **root causes of declining employee performance indexes** at INX Future Inc. and provide clear indicators for non-performing employees without negatively impacting overall employee morale or the company's reputation as a top employer.

- **Identifying patterns of attrition** within the employee data could be essential in understanding the underlying causes of the performance issues and developing strategies to mitigate them.

#### **IMPORTING THE LIBRARIES**

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

#### **IMPORTING THE DATASET**

In [4]:
!pip install xlrd>=2.0.1

In [5]:
data= pd.read_excel(r"E:\Employee_Performance_Analysis\INX_Future_Inc_Employee_Performance_CDS_Project2_Data_V1.8.xls")
data

EmpNumber  Age  Gender EducationBackground MaritalStatus  \
0     E1001000   32    Male           Marketing        Single   
1     E1001006   47    Male           Marketing        Single   
2     E1001007   40    Male       Life Sciences       Married   
3     E1001009   41    Male     Human Resources      Divorced   
4     E1001010   60    Male           Marketing        Single   
...        ...  ...     ...                 ...           ...   
1195   E100992   27  Female             Medical      Divorced   
1196   E100993   37    Male       Life Sciences        Single   
1197   E100994   50    Male             Medical       Married   
1198   E100995   34  Female             Medical        Single   
1199   E100998   24  Female       Life Sciences        Single   

        EmpDepartment        EmpJobRole BusinessTravelFrequency  \
0               Sales   Sales Executive           Travel_Rarely   
1               Sales   Sales Executive           Travel_Rarely   
2               Sales   Sales Executive       Travel_Frequently   
3     Human Resources           Manager           Travel_Rarely   
4               Sales   Sales Executive           Travel_Rarely   
...               ...               ...                     ...   
1195            Sales   Sales Executive       Travel_Frequently   
1196      Development  Senior Developer           Travel_Rarely   
1197      Development  Senior Developer           Travel_Rarely   
1198     Data Science    Data Scientist           Travel_Rarely   
1199            Sales   Sales Executive           Travel_Rarely   

      DistanceFromHome  EmpEducationLevel  ...  EmpRelationshipSatisfaction  \
0                   10                  3  ...                            4   
1                   14                  4  ...                            4   
2                    5                  4  ...                            3   
3                   10                  4  ...                            2   
4                   16                  4  ...                            4   
...                ...                ...  ...                          ...   
1195                 3                  1  ...                            2   
1196                10                  2  ...                            1   
1197                28                  1  ...                            3   
1198                 9                  3  ...                            2   
1199                 3                  2  ...                            1   

      TotalWorkExperienceInYears  TrainingTimesLastYear  EmpWorkLifeBalance  \
0                             10                      2                   2   
1                             20                      2                   3   
2                             20                      2                   3   
3                             23                      2                   2   
4                             10                      1                   3   
...                          ...                    ...                 ...   
1195                           6                      3                   3   
1196                           4                      2                   3   
1197                          20                      3                   3   
1198                           9                      3                   4   
1199                           4                      3                   3   

      ExperienceYearsAtThisCompany  ExperienceYearsInCurrentRole  \
0                               10                             7   
1                                7                             7   
2                               18                            13   
3                               21                             6   
4                                2                             2   
...                            ...                           ...   
1195                             6           

#### **DOMAIN ANALYSIS**

1. ***EmpNumber:*** Unique identifier for each employee in the dataset.

2. ***Age:*** Age of the employee, providing insight into workforce demographics and potential correlations with attrition.

3. ***Gender:*** Gender of the employee, which may impact workplace dynamics and attrition patterns.

4. ***EducationBackground:*** The educational background of the employee, influencing skillset and career trajectory.

5. ***MaritalStatus:*** Marital status of the employee, potentially affecting work-life balance and job satisfaction.

6. ***EmpDepartment:*** Department in which the employee works, indicating job role and organizational structure.

7. ***EmpJobRole:*** Specific job role of the employee within their department, reflecting responsibilities and career path.

8. ***BusinessTravelFrequency:*** Frequency of business travel for the employee, impacting lifestyle and job satisfaction.

9. ***DistanceFromHome:*** Distance of employee's residence from the workplace, influencing commuting stress and retention.

10. ***EmpEducationLevel:*** Level of education attained by the employee, reflecting qualifications and potential for advancement.

11. ***EmpEnvironmentSatisfaction:*** Employee satisfaction with the work environment, affecting morale and turnover.

12. ***EmpHourlyRate:*** Hourly wage of the employee, a factor in compensation satisfaction and retention.

13. ***EmpJobInvolvement:*** Level of involvement and engagement in the job role, affecting performance and attrition risk.

14. ***EmpJobLevel:*** Level of hierarchy within the organization, indicating seniority and career progression.

15. ***EmpJobSatisfaction:*** Satisfaction level with the job role, impacting employee morale and retention.

16. ***NumCompaniesWorked:*** Number of companies the employee has previously worked for, indicating job stability and turnover risk.

17. ***OverTime:*** Whether the employee works overtime, influencing work-life balance and burnout.

18. ***EmpLastSalaryHikePercent:*** Percentage of the employee's last salary hike, affecting compensation satisfaction and retention.

19. ***EmpRelationshipSatisfaction:*** Satisfaction with relationships at work, influencing job satisfaction and likelihood of turnover.

20. ***TotalWorkExperienceInYears:*** Total work experience of the employee, influencing skill level and career trajectory.

21. ***TrainingTimesLastYear:*** Number of training sessions attended by the employee last year, indicating investment in skill development and career growth.

22. ***EmpWorkLifeBalance:*** Employee's perceived balance between work and personal life, affecting job satisfaction and retention.

23. ***ExperienceYearsAtThisCompany:*** Years of experience at the current company, indicating loyalty and potential for promotion.

24. ***ExperienceYearsInCurrentRole:*** Years of experience in the current job role, influencing expertise and potential for advancement.

25. ***YearsSinceLastPromotion:*** Time since the employee's last promotion, impacting career progression and job satisfaction.

26. ***YearsWithCurrManager:*** Years of tenure with the current manager, affecting job satisfaction and retention.

27. ***Attrition:*** Target variable indicating whether the employee has left the company, essential for attrition prediction modeling.

28. ***PerformanceRating:*** Performance rating assigned to the employee, influencing career development and potential for retention.

#### **BASIC CHECKS**

In [9]:
data.shape

(1200, 28)

In [6]:
# last five rows
data.tail()

EmpNumber  Age  Gender EducationBackground MaritalStatus EmpDepartment  \
1195   E100992   27  Female             Medical      Divorced         Sales   
1196   E100993   37    Male       Life Sciences        Single   Development   
1197   E100994   50    Male             Medical       Married   Development   
1198   E100995   34  Female             Medical        Single  Data Science   
1199   E100998   24  Female       Life Sciences        Single         Sales   

            EmpJobRole BusinessTravelFrequency  DistanceFromHome  \
1195   Sales Executive       Travel_Frequently                 3   
1196  Senior Developer           Travel_Rarely                10   
1197  Senior Developer           Travel_Rarely                28   
1198    Data Scientist           Travel_Rarely                 9   
1199   Sales Executive           Travel_Rarely                 3   

      EmpEducationLevel  ...  EmpRelationshipSatisfaction  \
1195                  1  ...                            2   
1196                  2  ...                            1   
1197                  1  ...                            3   
1198                  3  ...                            2   
1199                  2  ...                            1   

      TotalWorkExperienceInYears  TrainingTimesLastYear  EmpWorkLifeBalance  \
1195                           6                      3                   3   
1196                           4                      2                   3   
1197                          20                      3                   3   
1198                           9                      3                   4   
1199                           4                      3                   3   

      ExperienceYearsAtThisCompany  ExperienceYearsInCurrentRole  \
1195                             6                             5   
1196                             1                             0   
1197                            20                             8   
1198                             8                             7   
1199                             2                             2   

     YearsSinceLastPromotion  YearsWithCurrManager  Attrition  \
1195                       0                     4         No   
1196                       0                     0         No   
1197                       3                     8         No   
1198                       7                     7         No   
1199                       2                     0        Yes   

      PerformanceRating  
1195                  4  
1196                  3  
1197                  3  
1198                  3  
1199                  2  

[5 rows x 28 columns]

In [7]:
# columns
data.columns

Index(['EmpNumber', 'Age', 'Gender', 'EducationBackground', 'MaritalStatus',
       'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency',
       'DistanceFromHome', 'EmpEducationLevel', 'EmpEnvironmentSatisfaction',
       'EmpHourlyRate', 'EmpJobInvolvement', 'EmpJobLevel',
       'EmpJobSatisfaction', 'NumCompaniesWorked', 'OverTime',
       'EmpLastSalaryHikePercent', 'EmpRelationshipSatisfaction',
       'TotalWorkExperienceInYears', 'TrainingTimesLastYear',
       'EmpWorkLifeBalance', 'ExperienceYearsAtThisCompany',
       'ExperienceYearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager', 'Attrition', 'PerformanceRating'],
      dtype='object')

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 28 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   EmpNumber                     1200 non-null   object
 1   Age                           1200 non-null   int64 
 2   Gender                        1200 non-null   object
 3   EducationBackground           1200 non-null   object
 4   MaritalStatus                 1200 non-null   object
 5   EmpDepartment                 1200 non-null   object
 6   EmpJobRole                    1200 non-null   object
 7   BusinessTravelFrequency       1200 non-null   object
 8   DistanceFromHome              1200 non-null   int64 
 9   EmpEducationLevel             1200 non-null   int64 
 10  EmpEnvironmentSatisfaction    1200 non-null   int64 
 11  EmpHourlyRate                 1200 non-null   int64 
 12  EmpJobInvolvement             1200 non-null   int64 
 13  EmpJobLevel       

##### Insights:
- There are 19 features with int64 datatype and the rest of the 9 fetures are of object datatype.
- There are no null values in the records.

In [11]:
# viewing the statisticl information
data.describe().T.style.background_gradient(cmap= 'cool')

The table provides statistical summaries for various employee attributes. Here's a concise analysis of each:

1. **Age:**
   - The average age of employees is approximately 36.92 years, with a standard deviation of about 9.09 years. The age distribution is fairly spread out, ***ranging from 18 to 60 years.***

2. **DistanceFromHome:**
   - On average, employees live around 9.17 miles away from their workplace, with a standard deviation of about 8.18 miles. The ***distance varies widely, from 1 to 29 miles***.

3. **EmpEducationLevel:**
   - The average education level is about 2.89, with a standard deviation of approximately 1.04. The majority of employees have an education level between 1 and 5.

4. **EmpEnvironmentSatisfaction:**
   - The average satisfaction level is around 2.72, with a standard deviation of about 1.09. The majority of employees rate their environment satisfaction between 1 and 4.

5. **EmpHourlyRate:**
   - The average hourly rate is approximately $65.98, with a standard deviation of about $20.21. Hourly rates ***range from $30 to $100.***

6. **EmpJobInvolvement:**
   - On average, employees show a job involvement level of about 2.73, with a standard deviation of approximately 0.71. Most employees rate their job involvement between 1 and 4.

7. **EmpJobLevel:**
   - The average job level is around 2.07, with a standard deviation of about 1.11. Job levels range from 1 to 5, indicating a diverse workforce.

8. **EmpJobSatisfaction:**
   - The average job satisfaction level is about 2.73, with a standard deviation of around 1.10. The majority of employees rate their job satisfaction between 1 and 4.

9. **NumCompaniesWorked:**
   - On average, employees have worked for around 2.67 companies, with a standard deviation of approximately 2.47. ***The number of companies worked for ranges from 0 to 9.***

10. **EmpLastSalaryHikePercent:**
    - The average salary hike percentage is approximately 15.22%, with a standard deviation of about 3.63%. ***Salary hike percentages vary from 11% to 25%.***

11. **EmpRelationshipSatisfaction:**
    - On average, employees report a relationship satisfaction level of about 2.72, with a standard deviation of around 1.08. The majority of employees rate their relationship satisfaction between 1 and 4.

12. **TotalWorkExperienceInYears:**
    - The average total work experience is approximately 11.33 years, with a standard deviation of about 7.80 years. ***Total work experience ranges from 0 to 40 years, indicating a diverse workforce in terms of experience.***

13. **TrainingTimesLastYear:**
    -  On average, employees underwent training around 2.79 times last year, with a standard deviation of approximately 1.26 times. The number of training sessions varies from 0 to 6.

14. **EmpWorkLifeBalance:**
    - The average work-life balance rating is about 2.74, with a standard deviation of around 0.70. Most employees rate their work-life balance between 1 and 4.

15. **ExperienceYearsAtThisCompany:**
    - Employees have, on average, spent around 7.08 years at the current company, with a standard deviation of about 6.24 years. Experience years range from 0 to 40 years, indicating a diverse range of tenures.

These information can be helpful in arriving a comprehensive understanding of the demographic and professional characteristics of the workforce, which can inform HR policies and management decisions.

In [16]:
data.describe(include= 'O').T


count unique              top  freq
EmpNumber                1200   1200         E1001000     1
Gender                   1200      2             Male   725
EducationBackground      1200      6    Life Sciences   492
MaritalStatus            1200      3          Married   548
EmpDepartment            1200      6            Sales   373
EmpJobRole               1200     19  Sales Executive   270
BusinessTravelFrequency  1200      3    Travel_Rarely   846
OverTime                 1200      2               No   847
Attrition                1200      2               No  1022

##### Insights:
1. **Employee Number**:
   - All employees have a ***unique identification number.***
   - Each employee number appears only once in the dataset.

2. **Gender**:
   - Majority of the employees are male (725 out of 1200).
   - There is a ***gender imbalance*** in the workforce.

3. **Education Background**:
   - The most common educational background is in ***Life Sciences*** (492 employees).
   - Diverse educational backgrounds are represented.

4. **Marital Status**:
   - The majority of employees are ***married (548 out of 1200).***
   - Single and other marital statuses are less common.

5. **Employee Department**:
   - ***Sales is the largest department,*** with 373 employees.
   - The workforce is distributed across multiple departments.

6. **Employee Job Role**:
   - Sales Executive is the most common job role (270 employees).
   - A variety of job roles are present within the organization.

7. **Business Travel Frequency**:
   - Travel Rarely is the most common business travel frequency (846 employees).
   - Some employees may travel more frequently for work.

8. **Overtime**:
   - The ***majority of employees do not work overtime*** (847 out of 1200).
   - Work-life balance might be prioritized in this organization.

9. **Attrition**:
   - The majority of employees have not left the company (1022 out of 1200).
   - ***Attrition rate*** appears to be relatively low.

In [20]:
data.nunique().to_frame().T.style.background_gradient(cmap= 'cool')

##### Insights:
- **Numerical features**:
  -  **continuous_features:** 'EmpNumber', 'Age', 'DistanceFromHome', EmpHourlyRate', 'TotalWorkExperienceInYears', 'ExperienceYearsAtThiscompany'.
  -  **discrete_features:**  'EmpEducationLevel', 'EmpEnvironmentSatisfaction', 'EmpJobInvolvement', 'EmpJobLevel', 'EmpJobSatisfaction', 'NumCompaniesWorked', 'EmpLastSalaryHikePercent', 'EmpRelationshipSatisfaction', 'TrainingTimesLastYear', 'EmpWorkLifeBalance', 'ExperienceYearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager', 'PerformanceRating'.

- **Categorical features**:
  -  **binary_features** 'OverTime', 'Attrition'.
  -  **ordinal_features:** 'EmpEducationLevel'.
  -  **nominal_fetures:** 'Gender', 'EducationBackground', 'Maritalstatus', 'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency'.  